rfc/text/wamp-terminology.md

```diff
@@ -0,0 +1,75 @@
+|-----------------------------------|---------------------------------------------------------------------------------------------------------|
+| *User*                            | A person (or organization) running a WAMP *Client* or *Router* |
+| *Client*                          | A program run by a *User*, with application code using WAMP for application level communications        |
+| *Router*                          | A program run by a *User*, with middleware code using WAMP to provide application routing services      |
+| *Peer*                            | A WAMP *Client* or *Router*. An implementation might embed, provide or use both roles                   |
+| *Realm*                           | Isolated WAMP URI namespace serving as a routing and administrative domain, optionally protected by **AA**          |
+| *Transport*                       | A message-based, reliable, ordered, bidirectional (full-duplex) channel over which *Peers* communicate  |
+| *Connection*                      | An underlying entity (if any) carrying the *Transport*, e.g. a network connection, pipe, queue or such  |
+| *Session*                         | Transient conversation between a *Client* and a *Router*, within a *Realm* and over a *Transport*                |
+| *Message*                         | Indivisible unit of information transmitted between peers                                               |
+| *Serializer*                      | Encodes WAMP messages, with application payloads, into byte strings for transport         |
+
+
+### Authentication and Authorization (AA)
+
+{align="left"}
+| Term                               | Definition                                                                                                                     |
+|------------------------------------|--------------------------------------------------------------------------------------------------------------------------------|
+| *Authentication*                   | Establishes the identity of a *Session* within a *Realm*                                                                           |
+| *Principal*                        | The authenticated entity associated with a *Session* |
+| *Credentials*                      | The authentication information and secrets used during *Authentication*                                                        |
+| *Authorization*                    | A decision on permitting a *Principal* to perform a given *Action* on an *URI or URI pattern* |
+| *Access Control*                   | Policy for selective restriction of *Action*s on *URIs or URI patterns* performed by *Principal*s                             |
+| *Role-based Access Control (RBAC)* | An *Access Control* policy based on *Realm* (`realm`), *Principal*'s *Role* (`authrole`), *URI or URI pattern*, and *Action*   |
```

Comment:
If you don't mind, let's add also ABAC for the whole picture.

Attribute-based access control (ABAC) — Policy-based access control that defines an access control paradigm whereby a Principal's authorization to perform a set of operations is determined by evaluating attributes associated with the Principal, URI, Action, and, in some cases, environment attributes.

Reply:
I do mind;) I fact:

"Unlike [role-based access control (RBAC)](https://en.wikipedia.org/wiki/RBAC), which defines roles that carry a specific set of privileges associated with them and to which subjects are assigned, ABAC can express complex rule sets that can evaluate many different attributes."

fwiw, I deliberately decided against that and base policies only on authrole on realm - this includes APIs for dynamic auuthentication/authorization.

IOW: -1 from my side

Reply:
Well, maybe it makes sense. 

But just one simple example: let's say your dynamic authentication/authorization component takes into account in addition to authid/authrole time of the day to permit actions only during working hours. This is ABAC and not RBAC.

Reply:
yes, that would be ABAC. and RBAC is a simple subset of ABAC. the name ABAC is slightly misleading IMO, as it really is much more than attributes of the principal, but

> subject's authorization to perform a set of operations is determined by evaluating attributes associated with the subject, object, requested operations, and, in some cases, environment attributes.

eg your example: the working hours is an attribute of the environment.

the complexity introduced is IMO unjustified for the gain, eg caching becomes pretty unclear how to do it safely